File: lib/genetic_algorithm/genomes/EphemeralKeyGenome.py
import random

from lib.genetic_algorithm.genomes.SingleKeyGenome import SingleKeyGenome


class EphemeralKeyGenome(SingleKeyGenome):
    """
    A genome that represents a certain key that was pressed for a number of frames
    """

    def __init__(self, key_input, frames):
        """
        :param key_input: <str> the input key to press
        :param frames: <int> the number of frames the actions will last
        """
        SingleKeyGenome.__init__(self, key_input)
        self.frames = frames

    def get_copy(self):
        """
        Returns a copy of the current genome
        :return: <EphemeralKeyGenome> a copy of the current genome
        """
        return EphemeralKeyGenome(self.key_input, self.frames)

    def get_frames(self):
        """
        getter
        """
        return self.frames

    def mutate_with_max_frame(self, max_frame):
        """
        Mutates the frame of the individual and changes it to a random frame from 1 to the max frame
        :param max_frame: <int> the max frame for the genome to mutate
        """
        self.frames = random.randint(1,max_frame)
        SingleKeyGenome.mutate(self)

File: love_ga_wrapper/love_ga_wrapper/lib/genetic_algorithms/genomes/EphemeralKeyGenome.lua
local class = require("love_ga_wrapper.lib.classes.class")
local extend = require("love_ga_wrapper.lib.classes.extend")
local NullGenome = require("love_ga_wrapper.lib.genetic_algorithms.genomes.NullGenome")
-------------------------------------------------------------

-- class: EphemeralKeyGenome
-- param: input:str -> the input key to press
-- param: frames:int(1,) -> the number of frames the actions will last
-- param: word_dict:list(str) -> the word dictionary with all the possible inputs
-- A genome that stores a certain key that was pressed for a number of frames
local EphemeralKeyGenome = extend(NullGenome, function(self, input, frames, word_dict)
    if frames < 1 then
      error("EphemeralKeyGenome's frame has to be an integer higher than 0")
    end
    
    self.input = input
    self.frame = frames
    self.word_dict = word_dict
end,

function(inputs, word_dict)
  return NullGenome.new()
end)

-- crossOver: EphemeralKeyGenome -> EphemeralKeyGenome
-- Throws an error, because individuals must implement this genome's crossover
function EphemeralKeyGenome.crossOver(self, genome)
  error("This genome cant do crossover with another genome, the crossover operation must be done between individuals.")
end

-- mutate: None -> None
-- Trows an error, because this genome has many possible mutation functions and individuals must implement a mutation
-- that decides which to use
function EphemeralKeyGenome.mutate(self)
  error("This genome has various mutation functions, thus the mutation operation must be done by individuals.")
end

-- dFrameMutation: int -> None
-- Varies the frame where the input was made
function EphemeralKeyGenome.dFrameMutation(self, df)
  local random_frame_delta = math.random(-df,df)
  self.frame = math.max(1, self.frame + random_frame_delta)
end

-- keyChangeMutation: None -> None
-- Changes the current input key to a random input
function EphemeralKeyGenome.keyChangeMutation(self)
  local word_dict_len = (# self.word_dict)
  local random_input = self.word_dict[math.random(1, word_dict_len)]
  self.input = random_input
end

-- mutateWithMaxFrame: int -> None
-- mutates the genome using an individual's largerst frame
-- performs a dFrame mutation and then it performs a keyChangeMutation
function EphemeralKeyGenome.mutateWithMaxFrame(self, max_frame)
    self:dFrameMutation(10)
    self:keyChangeMutation()
end

-- getter
function EphemeralKeyGenome.getInput(self)
    return self.input
end

-- getter
function EphemeralKeyGenome.getFrames(self)
    return self.frame
end

-- getter
function EphemeralKeyGenome.getWordDict(self)
    return self.word_dict
end

-- getter
function EphemeralKeyGenome.getInputs(self)
  local inputs = {}
  inputs[self.input] = true
  return inputs
end

return EphemeralKeyGenome